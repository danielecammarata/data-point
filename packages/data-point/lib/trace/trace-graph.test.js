/* eslint-env jest */

const { format } = require("util");
const set = require("lodash/set");

const TraceGraph = require("./trace-graph");

function createNode(index) {
  const time = index * 1e9;
  return {
    id: `id${index}`,
    time,
    timeStartNs: time,
    durationNs: time + 20000,
    parent: null,
    reducer: {
      id: `reducerId${index}`,
      name: `reducerName${index}`
    }
  };
}

function createTraceGraph() {
  const traceGraph = Array(4)
    .fill(null)
    .map((val, index) => {
      return createNode(index);
    });

  traceGraph[1].parent = traceGraph[0];
  traceGraph[2].parent = traceGraph[0];
  traceGraph[3].parent = traceGraph[2];

  return traceGraph;
}

function createGraph() {
  const traceGraph = createTraceGraph();
  const root = traceGraph[0];
  const graphAcc = { timeStartNs: root.timeStartNs, maxNestingLevel: 0 };
  TraceGraph.createTree(root, traceGraph, 0, graphAcc);
  return {
    root,
    graphAcc
  };
}

describe("createReducerSummary", () => {
  it("should create summary for non ReducerFunction", () => {
    const reducer = {
      id: "id",
      type: "ReducerObject",
      name: "name"
    };
    expect(TraceGraph.createReducerSummary(reducer)).toEqual({
      body: "",
      id: "id",
      name: "name",
      type: "ReducerObject"
    });
  });

  it("should create summary for reducer with no id or name", () => {
    const reducer = {
      type: "ReducerObject"
    };
    expect(TraceGraph.createReducerSummary(reducer)).toEqual({
      body: "",
      id: "",
      name: "",
      type: "ReducerObject"
    });
  });

  it("should create summary for ReducerFunction with named function", () => {
    const reducer = {
      id: "id",
      type: "ReducerFunction",
      name: "name",
      body: function namedFunction() {}
    };
    expect(TraceGraph.createReducerSummary(reducer)).toEqual({
      id: "id",
      name: reducer.body.name,
      type: "ReducerFunction",
      body: reducer.body.toString()
    });
  });

  it("should create summary for ReducerFunction with anonymous function", () => {
    // to create an anonymous
    const createFunction = () => () => {};
    const reducer = {
      id: "id",
      type: "ReducerFunction",
      name: "name",
      body: createFunction()
    };
    expect(TraceGraph.createReducerSummary(reducer)).toEqual({
      id: "id",
      name: "anonymous",
      type: "ReducerFunction",
      body: reducer.body.toString()
    });
  });
});

describe("createTraceNodeLabel", () => {
  it("should use name when available", () => {
    const node = {};
    set(node, "id", "nodeId");
    set(node, "reducer.id", "reducerId");
    set(node, "reducer.name", "reducerName");
    expect(TraceGraph.createTraceNodeLabel(node)).toEqual("reducerId:nodeId");
  });
  it("should not set name when not available", () => {
    const node = {};
    set(node, "id", "nodeId");
    set(node, "reducer.id", "reducerId");
    expect(TraceGraph.createTraceNodeLabel(node)).toEqual("reducerId:nodeId");
  });
  it("should use reducer type when id is not available", () => {
    const node = {};
    set(node, "id", "nodeId");
    set(node, "reducer.type", "reducerType");
    expect(TraceGraph.createTraceNodeLabel(node)).toEqual("reducerType:nodeId");
  });
});

describe("createTree", () => {
  it("should create parent child relationship", () => {
    const result = createGraph();
    expect(result.root).toMatchSnapshot();
  });

  it("should set maxNestingLevel", () => {
    const result = createGraph();
    expect(result.graphAcc.maxNestingLevel).toEqual(3);
  });
});

describe("logGraph", () => {
  let mockConsoleLog;
  afterEach(() => {
    mockConsoleLog.mockRestore();
  });

  it("should log the tree", () => {
    const logs = [];
    mockConsoleLog = jest
      .spyOn(console, "log")
      .mockImplementation((...args) => {
        logs.push(format(...args));
      });

    TraceGraph.logGraph(createGraph().root);
    const logOutput = logs.join("\n");
    expect(logOutput).toMatchSnapshot();
  });
});

describe("writeTraceGraph", () => {
  let mockDateNow;
  let mockWriteFileP;
  beforeEach(() => {
    mockDateNow = jest.spyOn(Date, "now").mockImplementation(() => {
      return 123;
    });
    mockWriteFileP = jest
      .spyOn(TraceGraph, "writeFileP")
      .mockImplementation(() => {
        return Promise.resolve(true);
      });
  });
  afterEach(() => {
    mockDateNow.mockRestore();
    mockWriteFileP.mockRestore();
  });
  it("should write graph to disk", async () => {
    const traceGraph = createTraceGraph();
    await TraceGraph.writeTraceGraph(traceGraph);
    expect(mockWriteFileP.mock.calls[0]).toMatchSnapshot();
  });
});
