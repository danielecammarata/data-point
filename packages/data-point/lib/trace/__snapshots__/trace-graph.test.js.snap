// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTree should create parent child relationship 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "durationMs": 1000.02,
      "durationNs": 1000020000,
      "id": "id1",
      "label": "reducerId1:id1",
      "nestingLevel": 1,
      "parent": [Circular],
      "reducer": Object {
        "id": "reducerId1",
        "name": "reducerName1",
      },
      "reducerSummary": Object {
        "body": "",
        "id": "reducerId1",
        "name": "reducerName1",
        "type": undefined,
      },
      "time": 1000000000,
      "timeStartNs": 1000000000,
      "timelineStartMs": 1000,
      "timelineStartNs": 1000000000,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "durationMs": 3000.02,
          "durationNs": 3000020000,
          "id": "id3",
          "label": "reducerId3:id3",
          "nestingLevel": 2,
          "parent": [Circular],
          "reducer": Object {
            "id": "reducerId3",
            "name": "reducerName3",
          },
          "reducerSummary": Object {
            "body": "",
            "id": "reducerId3",
            "name": "reducerName3",
            "type": undefined,
          },
          "time": 3000000000,
          "timeStartNs": 3000000000,
          "timelineStartMs": 3000,
          "timelineStartNs": 3000000000,
        },
      ],
      "durationMs": 2000.02,
      "durationNs": 2000020000,
      "id": "id2",
      "label": "reducerId2:id2",
      "nestingLevel": 1,
      "parent": [Circular],
      "reducer": Object {
        "id": "reducerId2",
        "name": "reducerName2",
      },
      "reducerSummary": Object {
        "body": "",
        "id": "reducerId2",
        "name": "reducerName2",
        "type": undefined,
      },
      "time": 2000000000,
      "timeStartNs": 2000000000,
      "timelineStartMs": 2000,
      "timelineStartNs": 2000000000,
    },
  ],
  "durationMs": 0.02,
  "durationNs": 20000,
  "id": "id0",
  "label": "reducerId0:id0",
  "nestingLevel": 0,
  "parent": null,
  "reducer": Object {
    "id": "reducerId0",
    "name": "reducerName0",
  },
  "reducerSummary": Object {
    "body": "",
    "id": "reducerId0",
    "name": "reducerName0",
    "type": undefined,
  },
  "time": 0,
  "timeStartNs": 0,
  "timelineStartMs": 0,
  "timelineStartNs": 0,
}
`;

exports[`logGraph should log the tree 1`] = `
"0 reducerId0:id0 S: 0.000ms T: 0.000ms
  1 reducerId1:id1 S: 1.000ms T: 1.000ms
  1 reducerId2:id2 S: 2.000ms T: 2.000ms
    2 reducerId3:id3 S: 3.000ms T: 3.000ms"
`;

exports[`writeTraceGraph should write graph to disk 1`] = `
Array [
  "data-point-trace-123.json",
  "{
  \\"id\\": \\"id0\\",
  \\"time\\": 0,
  \\"timeStartNs\\": 0,
  \\"durationNs\\": 20000,
  \\"parent\\": null,
  \\"reducer\\": {
    \\"id\\": \\"reducerId0\\",
    \\"name\\": \\"reducerName0\\"
  },
  \\"durationMs\\": 0.02,
  \\"timelineStartNs\\": 0,
  \\"timelineStartMs\\": 0,
  \\"nestingLevel\\": 0,
  \\"label\\": \\"reducerId0:id0\\",
  \\"reducerSummary\\": {
    \\"id\\": \\"reducerId0\\",
    \\"name\\": \\"reducerName0\\",
    \\"body\\": \\"\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"id1\\",
      \\"time\\": 1000000000,
      \\"timeStartNs\\": 1000000000,
      \\"durationNs\\": 1000020000,
      \\"parent\\": \\"[Circular ~]\\",
      \\"reducer\\": {
        \\"id\\": \\"reducerId1\\",
        \\"name\\": \\"reducerName1\\"
      },
      \\"durationMs\\": 1000.02,
      \\"timelineStartNs\\": 1000000000,
      \\"timelineStartMs\\": 1000,
      \\"nestingLevel\\": 1,
      \\"label\\": \\"reducerId1:id1\\",
      \\"reducerSummary\\": {
        \\"id\\": \\"reducerId1\\",
        \\"name\\": \\"reducerName1\\",
        \\"body\\": \\"\\"
      },
      \\"children\\": []
    },
    {
      \\"id\\": \\"id2\\",
      \\"time\\": 2000000000,
      \\"timeStartNs\\": 2000000000,
      \\"durationNs\\": 2000020000,
      \\"parent\\": \\"[Circular ~]\\",
      \\"reducer\\": {
        \\"id\\": \\"reducerId2\\",
        \\"name\\": \\"reducerName2\\"
      },
      \\"durationMs\\": 2000.02,
      \\"timelineStartNs\\": 2000000000,
      \\"timelineStartMs\\": 2000,
      \\"nestingLevel\\": 1,
      \\"label\\": \\"reducerId2:id2\\",
      \\"reducerSummary\\": {
        \\"id\\": \\"reducerId2\\",
        \\"name\\": \\"reducerName2\\",
        \\"body\\": \\"\\"
      },
      \\"children\\": [
        {
          \\"id\\": \\"id3\\",
          \\"time\\": 3000000000,
          \\"timeStartNs\\": 3000000000,
          \\"durationNs\\": 3000020000,
          \\"parent\\": \\"[Circular ~.children.1]\\",
          \\"reducer\\": {
            \\"id\\": \\"reducerId3\\",
            \\"name\\": \\"reducerName3\\"
          },
          \\"durationMs\\": 3000.02,
          \\"timelineStartNs\\": 3000000000,
          \\"timelineStartMs\\": 3000,
          \\"nestingLevel\\": 2,
          \\"label\\": \\"reducerId3:id3\\",
          \\"reducerSummary\\": {
            \\"id\\": \\"reducerId3\\",
            \\"name\\": \\"reducerName3\\",
            \\"body\\": \\"\\"
          },
          \\"children\\": []
        }
      ]
    }
  ],
  \\"maxNestingLevel\\": 3
}",
  "utf8",
]
`;
